### PR TITLE
fix: let etcd v3 `read_watch` handle the case where a chunk contains partial event or multiple events

### LINKED ISSUE #153: bug: watch responses may be incorrectly parsed in the case of ETCD being proxied by a load balancer like nginx
We have encountered some problems in using apisix, because our etcd uses nginx as a proxy. When etcd returns event responses as http chunk after passing through nginx proxy, the following two situations may occur: 
1. the data of an event is larger than the `proxy_buffer_size` of nginx, the origin chunk will be split into two chunks by nginx and sent to apisix, resulting in the `body_reader()` getting an incomplete json, and then the parsing will fail.
2. two consecutive events are merged into one chunk by nginx and sent to the client, which will also cause apisix to fail to parse the json.

https://github.com/api7/lua-resty-etcd/blob/f75f4c551b048e00233df38ab725b46218b35de9/lib/resty/etcd/v3.lua#L690-L703

I think the root cause of the above problem is that apisix assumes that each chunk in the watch response is a complete event, in fact, a chunk may be a partial event or multiple events, so we should take these into account.

Since the raw data returned by etcd carries a newline at the end of each complete event, I think we can use newlines as a event splitter to solve the above problem.

### PR DESCRIPTION
**Pull request by nic-6443**
Try to fix #153, we have verified in our environment that can solve that problem, but the test cases still need time to complete.

### CODE REVIEW

**Review by membphis — COMMENTED** [review]
missing test cases

**Comment by nic-6443** [comment]
We have a tricky problem, our CI includes a test of ETCD version 3.3.0, but this version of ETCD has a bug in the [grpc-gateway](https://github.com/grpc-ecosystem/grpc-gateway) dependency it relies on to implement HTTP API, which causes every event json to not have a trailing newline, which is the key I use to split the event. 
This means the code in this PR does not work properly on version 3.3.0 of ETCD.

References:
* ETCD v3.3.0 changelog: https://github.com/etcd-io/etcd/blob/v3.3.0/CHANGELOG.md#changedbreaking-changes
> Upgrade [github.com/grpc-ecosystem/grpc-gateway](https://github.com/grpc-ecosystem/grpc-gateway/releases) v1.2.2 to v1.3.0.
* The PR that introduced the bug in grpc-gateway: https://github.com/grpc-ecosystem/grpc-gateway/pull/479, it's include in grpc-gateway v1.3.0, so ETCD v3.3.0 been infected.
* After the introduction of the bug, the grpc-gateway community quickly had user feedback on the issue (https://github.com/grpc-ecosystem/grpc-gateway/issues/495), and the bug was fixed at https://github.com/grpc-ecosystem/grpc-gateway/pull/497, that adds back trailing new line in default
* [The list of features of grpc-gateway](https://github.com/grpc-ecosystem/grpc-gateway#features) states that newlines are used as separators for the stream api
> Mapping streaming APIs to newline-delimited JSON streams.


To summarize, ETCD uses grpc-gateway to implement the HTTP API, and grpc-gateway explicitly states that it will use newlines as a separator for stream-type grpc responses, but because of a bug introduced in the historical version, it does not pass the test cases of v3.3.0.

**Comment by nic-6443** [comment]
@tokers @starsz  @membphis @spacewander Any idea?

**Comment by membphis** [comment]
we can skip some test case for etcd `v3.3.0`

https://github.com/apache/apisix/blob/a11b3cc9438ef1d0074bb9d95f8a83521e6d364e/t/node/not-exist-upstream.t#L19

**Comment by nic-6443** [comment]
> We can drop the support of etcd < 3.4

Got it, thanks.

File: lib/resty/etcd/v3.lua [hunk 1]
```diff
@@ -688,43 +689,74 @@ local function request_chunk(self, method, path, opts, timeout)
...
+            end
+            if not chunk then
+                break
+            end
+
             if not body then
-                return nil, err
+                body = chunk
+            else
+                body = body .. chunk
```
Comment (spacewander) [inline lib/resty/etcd/v3.lua:706]: Could we use `table.concat` to avoid creating too much temp str?

File: lib/resty/etcd/v3.lua [hunk 2]
```diff
@@ -688,43 +689,74 @@ local function request_chunk(self, method, path, opts, timeout)
...
             end
-            if not utils.is_empty_str(body) then
+
+            if not utils.is_empty_str(chunk) and sub_str(chunk, -1) == "\n" then
```
Comment (spacewander) [inline lib/resty/etcd/v3.lua:709]: Use `string.byte` would be better

File: lib/resty/etcd/v3.lua (hunk 1 above)
Comment (nic-6443) [reply lib/resty/etcd/v3.lua]: OK

File: lib/resty/etcd/v3.lua (hunk 2 above)
Comment (nic-6443) [reply lib/resty/etcd/v3.lua]: Done

**Review by membphis — APPROVED** [review]

File: lib/resty/etcd/v3.lua [hunk 3]
```diff
@@ -688,43 +690,74 @@ local function request_chunk(self, method, path, opts, timeout)
...
+        local chunks, error = split(body, [[\n]], "jo")
+        if error then
+            return nil, "failed to split chunks: " .. error
+        end
+
+        local all_events = {}
+        for _, chunk in ipairs(chunks) do
+            body, err = decode_json(chunk)
+            if not body then
+                return nil, "failed to decode json body: " .. (err or " unkwon")
```
Comment (tokers) [inline lib/resty/etcd/v3.lua:729]:
```suggestion
                return nil, "failed to decode json body: " .. (err or " unknown")
```

File: lib/resty/etcd/v3.lua [hunk 4]
```diff
@@ -688,43 +690,74 @@ local function request_chunk(self, method, path, opts, timeout)
...
+
             if not body then
-                return nil, err
+                body = chunk
+            else
+                body = tab_concat({body, chunk})
             end
-            if not utils.is_empty_str(body) then
+
+            if not utils.is_empty_str(chunk) and str_byte(chunk, -1) == str_byte("\n") then
```
Comment (tokers) [inline lib/resty/etcd/v3.lua:710]: So if the response chunk doesn't contain the `\n` and we'll continue reading the next chunk? But if the connection is aborted / timed out this time, so the last chunk was missed?

File: lib/resty/etcd/v3.lua [hunk 5]
```diff
@@ -688,43 +690,74 @@ local function request_chunk(self, method, path, opts, timeout)
...
+            end
+            if not chunk then
+                break
+            end
+
             if not body then
-                return nil, err
+                body = chunk
+            else
+                body = tab_concat({body, chunk})
```
Comment (starsz) [inline lib/resty/etcd/v3.lua:707]: I think we can call tab_contat out of the while.

Comment (nic-6443) [reply lib/resty/etcd/v3.lua:707]: Not good, loop may execute more than twice.

Comment (membphis) [reply lib/resty/etcd/v3.lua:707]: this way is worse

Comment (starsz) [reply lib/resty/etcd/v3.lua:707]: > Not good, loop may execute more than twice.

I mean just tab_concat once to concat all chunk

Comment (nic-6443) [reply lib/resty/etcd/v3.lua:707]: Got it

Comment (nic-6443) [reply lib/resty/etcd/v3.lua:707]: I thought about it carefully, and since this concat branch will only be executed in rare cases(while one chunk larger  than nginx `proxy_buffer_size`), so I think it would be more appropriate to use a string directly, which avoid creating a table with only one element in most cases.  
@membphis @spacewander @starsz Do you agree with this conclusion?

Comment (tokers) [reply lib/resty/etcd/v3.lua]: I think just use `..` is enough.

Comment (spacewander) [reply lib/resty/etcd/v3.lua]: We can use `..` but add a comment to descript why?

Comment (nic-6443) [reply lib/resty/etcd/v3.lua]: ok

**Review by starsz — APPROVED** [review]

File: lib/resty/etcd/v3.lua [hunk 6]
```diff
@@ -688,43 +689,77 @@ local function request_chunk(self, method, path, opts, timeout)
...
-            body, err = res.body_reader()
+            local chunk, error = res.body_reader()
+            if error then
+                return nil, error
+            end
+            if not chunk then
+                break
+            end
+
             if not body then
```
Comment (spacewander) [inline lib/resty/etcd/v3.lua:703]: Look like this branch is not executed, as `body` has default value `""`?

Comment (nic-6443) [reply lib/resty/etcd/v3.lua:703]: yeah, my fault

File: t/v3/key.t [hunk 7]
```diff
@@ -512,3 +512,151 @@ passed
...
+
+            ngx.say('{"result":{"header":{"cluster_id":"14841639068965178418","member_id":"10276657743932975437","revision":"290","raft_term":"8"},"created":true}}')
+            ngx.flush()
+            ngx.sleep(0.1)
+
+            -- one http chunk contains multiple event response, note the new line at the end of first event response
+            -- key1 = /wdir/, value1 = bcd4
+            -- key2 = /wdir/a, value2 = bcd4a
+            ngx.say('{"result":{"header":{"cluster_id":"14841639068965178418","member_id":"10276657743932975437","revision":"292","raft_term":"8"},"events":[{"kv":{"key":"L3dkaXIv","create_revision":"31","mod_revision":"292","version":"22","value":"ImJjZDQi"}}]}}\n{"result":{"header":{"cluster_id":"14841639068965178418","member_id":"10276657743932975437","revision":"293","raft_term":"8"},"events":[{"kv":{"key":"L3dkaXIvYQ==","create_revision":"293","mod_revision":"293","version":"1","value":"ImJjZDRhIg=="}}]}}')
+            ngx.flush()
```
Comment (tokers) [inline t/v3/key.t:615]: Missing a case that we send two json strings in the same HTTP chunk?

File: lib/resty/etcd/v3.lua [hunk 8]
```diff
@@ -688,43 +689,77 @@ local function request_chunk(self, method, path, opts, timeout)
...
-            return nil, endpoint.http_host .. ": " .. body.error.http_status
+          return nil, nil
         end
 
-        if body.result and body.result.events then
-            for _, event in ipairs(body.result.events) do
-                if event.kv.value then   -- DELETE not have value
-                    event.kv.value = decode_base64(event.kv.value or "")
-                    event.kv.value = self.serializer.deserialize(event.kv.value)
+        local chunks, error = split(body, [[\n]], "jo")
```
[the remark this answers is not in the document]
Comment (membphis) [reply lib/resty/etcd/v3.lua:722]: nice catch

File: t/v3/key.t (hunk 7 above)
Comment (nic-6443) [reply t/v3/key.t:615]: Isn't it the current case?

File: lib/resty/etcd/v3.lua (hunk 8 above)
[the remark this answers is not in the document]
Comment (nic-6443) [reply lib/resty/etcd/v3.lua]: OK, but since `err` is already defined outside of `read_watch`, if I use `err`, our lint will complain about duplicate variable names, so I replace `error` with  `split_err`.

**Review by spacewander — APPROVED** [review]

File: t/v3/key.t (hunk 7 above)
Comment (tokers) [reply t/v3/key.t:615]: OK, I found it....

**Review by tokers — APPROVED** [review]
